# DAA 2022 S2 Assignment

Candidate number: 222408

## Purpose:
*Provide a clear and concise statement of the purpose of this notebook.*

*NB: In this notebook template, text and code cells have been created for step 1 and step 2/3/4 etc. You should replicate this structure for each individual step you undertake under each of the question headings. *


## Packages
This section imports the packages that are used in this notebook.

In [1]:
# Packages for data management.
import pandas as pd
import numpy as np

# Packages to be used in pre-processing the text.
import re
import nltk
nltk.download('stopwords') # For a standard list of stopwords.
nltk.download('wordnet') # For lemmatisation.
import pkg_resources
!pip install symspellpy
from symspellpy import SymSpell, Verbosity

# Packages for visualisation.
# from pprint import pprint
# import seaborn as sns
import matplotlib.pyplot as plt
from wordcloud import WordCloud
from collections import Counter

# For vectorisation using TF-IDF and t-SNE.
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.manifold import TSNE


# For vectorisation using BERT.
!pip install sentence_transformers
from sentence_transformers import SentenceTransformer


# For uploading files to Colab.
from google.colab import files

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 2.6 MB 4.3 MB/s 
     |████████████████████████████████| 125 kB 55.5 MB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 85 kB 2.1 MB/s 
     |████████████████████████████████| 4.7 MB 39.3 MB/s 
     |████████████████████████████████| 1.3 MB 19.5 MB/s 
     |████████████████████████████████| 120 kB 34.8 MB/s 
     |████████████████████████████████| 6.6 MB 42.8 MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125938 sha256=fa044c6355cb96477f049296415312cd7b6a8af517c4cb884b07c4a924674983
  Stored in directory: /root/.cache/pip/wheels/bf/06/fb/d59c1e5bd1dac7f6cf61ec0036cc3a10ab8fecaa6b2c3d3ee9
Successfully built sentence-transformers


## Functions
This section sets out user defined functions that are used in this notebook.

In [ ]:
# Function 1

# Function 2/3/4 etc.


## Import data
This section imports the data that is used in this notebook.

*You should include a reference to the dataset you have used here.*

In [2]:
from google.colab import drive
drive.mount('/content/gdrive/',force_remount=True)

# Specify the Google Drive folder that your datasets are saved in.
infolder = '/content/gdrive/My Drive/DAA datasets/'

# Specify the name of the dataset to be imported.
file = 'DAA 2022 S2 Assignment - Property Data FINAL.xlsx'

# Upload the required YELP file from your Google Drive folder.
dataset = pd.read_excel(infolder+file)

Mounted at /content/gdrive/


In [ ]:
# Create dataset from the excel import.
files.upload()

dataset = pd.read_excel('DAA 2022 S2 Assignment - Property Data FINAL.xlsx')

print(dataset.shape)

## Q3a - Examine the property dataset
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q3a Step 1 - Data structure
**Outline**

*Provide a clear and concise outline here about the EDA that you have conducted in the section below. This should include what you are doing and why.*

In [ ]:
# Check the size of the dataset.
print(dataset.shape)

# Print the first 5 rows of data.
print(dataset.head())


(49608, 14)
                     Address  Suburb  Postcode  Beds  Bathrooms  Carspaces  \
0     323/158-166 Day Street  Sydney      2000     2        2.0          1   
1  Level 69/115  Bathurst St  Sydney      2000     2        2.0          1   
2        1564/37 King Street  Sydney      2000     1        1.0          0   
3   4504/93 Liverpool Street  Sydney      2000     2        2.0          1   
4  209/298-304 Sussex Street  Sydney      2000     1        1.0          0   

  PropertyType   SoldDate                       SoldPrice FirstAdDate  \
0    Apartment 2021-04-01                         1188000  2020-10-30   
1    Apartment 2021-04-04  Range: $2,310,000 - $2,690,000  2020-11-13   
2    Apartment 2021-04-12                          790000  2021-03-19   
3    Apartment 2021-04-14                         1385000  2021-02-26   
4    Apartment 2021-04-19                          750000  2021-04-02   

               AdPrice                                          AdHeading  \
0  

In [ ]:
# Print the last 5 rows of data.
print(dataset.tail())

                      Address     Suburb  Postcode  Beds  Bathrooms  \
49603      12 Cuttriss Street  Inverloch      3996     4        2.0   
49604  Unit 1/26 Freda Street  Inverloch      3996     2        2.0   
49605         10 Ripple Drive  Inverloch      3996     4        2.0   
49606          3 Ripple Drive  Inverloch      3996     0        NaN   
49607      25 Paperbark Place  Inverloch      3996     4        2.0   

       Carspaces      PropertyType   SoldDate SoldPrice FirstAdDate  AdPrice  \
49603          0             House 2021-06-23   1110000  2021-05-07  1300000   
49604          1              Unit 2021-06-24    745000  2021-05-28   745000   
49605          0             House 2021-06-27    900000  2021-05-21   895000   
49606          0  Residential Land 2021-06-29    625000  2021-06-11   625000   
49607          4             House 2021-06-29   1310000  2021-02-26  1360000   

                                           AdHeading  \
49603  DOORS TO THE SEA WITH DEVELOP

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the EDA conducted above.*

### Q3a Step 2 - Types of data
**Outline**

Here we check for the datatypes because sometimes variables might be stored as not desirable types (e.g. sales price is stored as a string rather than an integer), then we have to convert those types before processing the data

In [ ]:
# Check the types of each feature and the response variable ('label').
print(dataset.dtypes)


Address                 object
Suburb                  object
Postcode                 int64
Beds                     int64
Bathrooms              float64
Carspaces                int64
PropertyType            object
SoldDate        datetime64[ns]
SoldPrice               object
FirstAdDate     datetime64[ns]
AdPrice                 object
AdHeading               object
AdBody                  object
AgencyName              object
dtype: object


**Summary**  
The data types for most variables make sense except the below which requires further cleaning: 
- 'Bathrooms': the data type should be integer 
- 'SoldPrice' and 'AdPrice' should be floating-point number


### Clean Propertype



In [ ]:
#### pivot table#### 

cleandata['StandardType'] = cleandata['PropertyType']

for i in range (0,len(cleandata)):
  if cleandata.loc[i,'StandardType'] in ['Unit','Flat','Studio','Serviced Apartment']:
    cleandata.loc[i,'StandardType'] = "Apartment"

  elif cleandata.loc[i,'StandardType'] in ['Cropping','Dairy','Farmlet','Horticulture','Livestock','Mixed Farming']:
    cleandata.loc[i,'StandardType'] = "Agriculture"
  
  elif cleandata.loc[i,'StandardType'] in ['Duplex/Semi-detached','Terrace','Townhouse','Villa']: 
    cleandata.loc[i,'StandardType'] ='Semi-detached'
  
  elif cleandata.loc[i,'StandardType'] in ['Alpine','Lifestyle','Other','Retirement Living','Viticulture/vineyard','Warehouse']:
    cleandata.loc[i,'StandardType'] ='To remove'

#check results 
pd.pivot_table(data=cleandata,
               index=['StandardType','PropertyType'], 
               values=['Address'],
               aggfunc = 'count')



In [ ]:
# drop the type 'To Remove'
cleandata = cleandata[cleandata["StandardType"].str.contains("To remove")==False].reset_index(drop=True)


### Check up postcode


FileNotFoundError: ignored

### Duplicates check + remove


In [3]:
# check duplicates 
duplicate_rows_dataset = dataset[dataset.duplicated()]
print("number of duplicate rows: ", duplicate_rows_dataset.shape)
duplicate_rows_dataset

#drop duplicates 
print(dataset.count()) # count the number of rows before dropping 
cleandata = dataset.drop_duplicates().reset_index(drop=True)
print(cleandata.count()) #count the number of rows again after dropping duplicates


number of duplicate rows:  (8, 14)
Address         49607
Suburb          49608
Postcode        49608
Beds            49608
Bathrooms       46980
Carspaces       49608
PropertyType    49608
SoldDate        49608
SoldPrice       49608
FirstAdDate     49608
AdPrice         49608
AdHeading       49608
AdBody          49608
AgencyName      49608
dtype: int64
Address         49599
Suburb          49600
Postcode        49600
Beds            49600
Bathrooms       46975
Carspaces       49600
PropertyType    49600
SoldDate        49600
SoldPrice       49600
FirstAdDate     49600
AdPrice         49600
AdHeading       49600
AdBody          49600
AgencyName      49600
dtype: int64


### Miss values check + remove

In [ ]:
#check Null values 
print(cleandata.isnull().sum())

Address            1
Suburb             0
Postcode           0
Beds               0
Bathrooms       2625
Carspaces          0
PropertyType       0
SoldDate           0
SoldPrice          0
FirstAdDate        0
AdPrice            0
AdHeading          0
AdBody             0
AgencyName         0
dtype: int64


**Summary** 

There's a reason in the above step while counting the number of rows where 'Address' had 49599 counts and 'Bathroom' had 46975 counts over 49600 rows. 

Further examination is required to determine how to deal with the NULL values.

In [ ]:
# check Null values - Address
# extract the row
index_add = cleandata[cleandata['Address'].isnull()].index.tolist()
index_add
cleandata.loc[index_add]

#replace the missing value with "Unknown"
cleandata.loc[index_add,"Address"] = "Unknown"
cleandata.loc[index_add]



Address        Suburb  Postcode  Beds  Bathrooms  Carspaces  \
41741  Unknown  Landsborough      3384     0        NaN          0   

      PropertyType   SoldDate SoldPrice FirstAdDate AdPrice  \
41741    Lifestyle 2021-06-11    210000  2021-05-28  210000   

                            AdHeading  \
41741  Pyrenees Weekender/Hobby Block   

                                                  AdBody  \
41741  What a great opportunity to secure an approx 5...   

                               AgencyName  
41741  Elders Real Estate Ararat - ARARAT

In [ ]:
# check Null values - Bathrooms
# extract the row
index_bath = cleandata[cleandata['Bathrooms'].isnull()].index.tolist()
# check number of rows
print(len(index_bath))


2625


In [ ]:
#check the first 5 headers 
cleandata.loc[index_bath].head()


Address            Suburb  Postcode  Beds  \
32       Lot 2006, TBA Byron Road        Leppington      2006     0   
837              167 Malabar Road      South Coogee      2034     0   
1854  Lot 306, St Columbans Green  North Turramurra      2074     0   
1855         Lot 312, Toms Pocket  North Turramurra      2074     0   
1867         Lot 404, Toms Pocket  North Turramurra      2074     0   

      Bathrooms  Carspaces      PropertyType   SoldDate SoldPrice FirstAdDate  \
32          NaN          0  Residential Land 2021-06-05    580000  2021-01-29   
837         NaN          0  Residential Land 2021-05-15   2175000  2021-04-30   
1854        NaN          0  Residential Land 2021-04-13   2400000  2020-12-04   
1855        NaN          0  Residential Land 2021-04-16   2385000  2020-12-11   
1867        NaN          0  Residential Land 2021-04-28   2448000  2021-04-23   

            AdPrice                                          AdHeading  \
32    Contact Agent       Land or House and Land Packages : Leppington   
837         Auction                Rare Site opportunity - Vacant Land   
1854        2245000    Prestige Corner Position in St Columbans Green    
1855  Contact Agent                   Superb Vacant St Columbans Land    
1867  Contact Agent  Pristine 828sqm Land Lot at St Columbans New R...   

                                                 AdBody  \
32    Lot 2006, TBA, Byron Road Leppington.<br/><br/...   
837   A rare opportunity in a Coastal location, this...   
1854  935sqm Torrens Title corner location in the pr...   
1855  EXCLUSIVE . INCLUSIVE . ENCHANTING.<br/><br/>F...   
1867  Lot 404 Toms Pocket is one of ten level land l...   

                                             AgencyName  
32                        Multi Dynamic Auburn - AUBURN  
837   Ray White - Ray White Surry Hills | Alexandria...  
1854                                  Proland Marketing  
1855                                  Proland Marketing  
1867                                  Proland Marketing

In [ ]:
#the PropertyType is "Reinsidential Land" for the first 5 rows with Null Bathrooms 
output = pd.pivot_table(data=cleandata.loc[index_bath], 
                        index=['PropertyType'], 
                        aggfunc = 'count')

print(output)

# Comments: based on the Property Type, it makes sense to have 0 bathrooms based on the domain knowledge. 
# Suggest placing NULL with "0" 


                      AdBody  AdHeading  AdPrice  Address  AgencyName  \
PropertyType                                                            
Cropping                  24         24       24       24          24   
Farmlet                    4          4        4        4           4   
Horticulture               4          4        4        4           4   
Lifestyle                109        109      109      109         109   
Livestock                 24         24       24       24          24   
Mixed Farming             21         21       21       21          21   
Other                     75         75       75       75          75   
Residential Land        2363       2363     2363     2363        2363   
Viticulture/vineyard       1          1        1        1           1   

                      Bathrooms  Beds  Carspaces  FirstAdDate  Postcode  \
PropertyType                                                              
Cropping                      0    24         

In [ ]:
#placing NULL with "0"
cleandata.loc[index_bath,'Bathrooms'] = 0
cleandata.loc[index_bath].head()


In [ ]:
#final check nulls 
print(cleandata.isnull().values.any())

False


In [ ]:
#make integer 
cleandata = cleandata.astype({'Bathrooms':'int'})

#final check types 
print(cleandata.dtypes)

### Ad date < Sold date

In [ ]:
# Sold date should be on or after Ad date 
identifier = cleandata[cleandata['SoldDate'] < cleandata['FirstAdDate']].index.tolist()
len(identifier)
errorset = cleandata.loc[identifier]
errorset['diffindate']= errorset['SoldDate'] - errorset['FirstAdDate']
errorset['diffindate'].describe()
plt.hist(errorset['diffindate'].dt.days)




In [ ]:

#assume swapped -> swap 
cleandata.loc[identifier,'SoldDate'] = errorset['FirstAdDate']
cleandata.loc[identifier,'FirstAdDate'] = errorset['SoldDate']

#check
print(cleandata[cleandata['SoldDate'] < cleandata['FirstAdDate']].values.any()) 

False


### Sold price remove the range

In [ ]:
# SoldPrice's type is "object", which should be floating numbers. Investigate 
cleandata = dataset.drop_duplicates().reset_index(drop=True)

cleandata
import statistics

for i in range(0,len(cleandata)):
  if not isinstance(cleandata.loc[i,'SoldPrice'],int): 
   templist = cleandata.loc[i,'SoldPrice'].replace('-', '$').replace(',','').split("$")
   templist = (int(templist[1]),int(templist[3]))
   aveprice = statistics.mean(templist)
   cleandata.loc[i,'SoldPrice'] = aveprice

#make SoldPrice  
cleandata = cleandata.astype({'SoldPrice':'int'})

#check types 
cleandata.dtypes


### Ad Price 

In [ ]:
# upon rough checking, quite a lot of entries just have 'Auction' or 'Contact Agent'
# instead of listing a price 

#check those that contain an indication 

cleandata['AdPrice'].str.contains('$')
from datetime import datetime
count = 0 
for i in range(0,len(cleandata)):
  if isinstance(cleandata.loc[i,'AdPrice'],int): 
    count = count +1 
  if isinstance(cleandata.loc[i,'AdPrice'],datetime): 
    continue
  elif '$' in cleandata.loc[i,'AdPrice']:
    count = count + 1 

print(count)


TypeError: ignored

**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q3b - Data cleaning and pre-processing
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q3b Step 1
**Outline**

*Provide a clear and concise outline here about the data cleaning or pre-processing step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q3b Step 1


**Summary**  
*Provide a clear and concise summary here of your conclusions based on the data cleaning or pre-processing output. Is the output what was expected? If not, why not?*

### Q3b Step 2/3/4 etc.
**Outline**

*Provide a clear and concise outline here about the data cleaning or pre-processing step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q3b Step 2/3/4 etc.

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the data cleaning or pre-processing. Is the output what was expected? If not, why not?*

**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q4 - Calculate vectorised features
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q4 Step 1
**Outline**

*Provide a clear and concise outline here about the NLP step that you have undertaken in the section below. This should include what you are doing and why.*



In [ ]:
# Q4 Step 1

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the NLP step undertaken above. Is the output what was expected? If not, why not?*

### Q4 Step 2/3/4 etc.

**Outline**

*Provide a clear and concise outline here about the NLP step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q4 Step 2/3/4 etc.

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the NLP step undertaken above. Is the output what was expected? If not, why not?*

**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q5a - Calculate a response variable
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q5a Step 1
**Outline**

*Provide a clear and concise outline here about the step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5a Step 1

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the step undertaken above. Is the output what was expected? If not, why not?*

### Q5a Step 2/3/4 etc.
**Outline**

*Provide a clear and concise outline here about the step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5a Step 2/3/4 etc.

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the step undertaken above. Is the output what was expected? If not, why not?*

**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q5b - Construct an additional feature
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q5b Step 1
**Outline**

*Provide a clear and concise outline here about the step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5b Step 1

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the step undertaken above. Is the output what was expected? If not, why not?*

### Q5b Step 2/3/4 etc.
**Outline**

*Provide a clear and concise outline here about the step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5b Step 2/3/4 etc.

**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the step undertaken above. Is the output what was expected? If not, why not?*

**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q5c - Construct a classification model
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q5c Step 1
**Outline**

*Provide a clear and concise outline here about the model building step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5c Step 1


**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the model building step undertaken above. Is the output what was expected? If not, why not?*

*If this step involved running a new iteration of the model, you should clearly and concisely outline what the iteration hoped to fix, what changes were made to the model, and whether the iteration fixed the problem.*


### Q5c Step 2/3/4 etc.
**Outline**

*Provide a clear and concise outline here about the model building step that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5c Step 2/3/4 etc.


**Summary**  
*Provide a clear and concise summary here of your conclusions based on the output of the model building step undertaken above. Is the output what was expected? If not, why not?*

*If this step involved running a new iteration of the model, you should clearly and concisely outline what the iteration hoped to fix, what changes were made to the model, and whether the iteration fixed the problem.*


**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*

## Q5d - Evaluate model predictions
### Purpose:
*Provide a clear and concise statement of the purpose of this section.*

### References:
*Outline any references you have used in this section of your notebook here.*

### Q5d Step 1
**Outline**

*Provide a clear and concise outline here about the model evaluation that you have undertaken in the section below. This should include what you are doing and why.*


In [ ]:
# Q5d Step 1


**Summary**  
*Provide a clear and concise summary here of your conclusions based on the model evaluation undertaken above. Is the output what was expected? If not, why not?*


### Q5d Step 2/3/4 etc.
**Outline**

*Provide a clear and concise outline here about the model evaluation that you have undertaken in the section below. This should include what you are doing and why.*

In [ ]:
# Q5d Step 2/3/4 etc.


**Summary**  
*Provide a clear and concise summary here of your conclusions based on the model evaluation undertaken above. Is the output what was expected? If not, why not?*


**Overall summary**  
*Provide a clear and concise overall summary here of your conclusions from this section of the notebook.*